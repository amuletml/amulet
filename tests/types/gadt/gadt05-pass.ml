external val (+) : int -> int -> int = ""

type t 't 'a =
  | T : 'a -> t () 'a

let foo : forall 'a 't. ('a -> 'a) -> t 't 'a -> t 't 'a =
  fun f (T x) -> T (f x)

let bar : forall 't. t 't int -> t 't int =
  fun t -> match t with
  | T _ -> foo (+ 1) t
