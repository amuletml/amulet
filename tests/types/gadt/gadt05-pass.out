+ : int -> int -> int
t : type -> type -> type
T : Infer{'bg : type}. Spec{'t : 'bg}. Spec{'a : type}. ('t ~ unit) âŠƒ 'a -> t 't 'a
foo : Spec{'a : type}. Spec{'t : type}. ('a -> 'a) -> t 't 'a -> t 't 'a
bar : Spec{'t : type}. t 't int -> t 't int
